tools/cbmc/README.md

```diff
@@ -13,16 +13,61 @@ analysis tool
 how to run the proofs on your local clone of a:FR.
 
 
-Prerequisites
--------------
+Bulding and running proofs
+--------------------------
+
+For historical reasons, some of the proofs are built and run using CMake
+and CTest. Others use a custom python-based build system. New proofs
+should use CMake. This README describes how to build and run both kinds
+of proof.
+
+
+CMake-based build
+-----------------
+
+Follow the CBMC installation instructions below.
+
+Suppose that the amazon-freertos source tree is located at
+`~/src/amazon-freertos` and you wish to build the proofs into
+`~/build/amazon-freertos`. The following three commands build and run
+the proofs:
+
+```sh
+cmake -S~/src/amazon-freertos -B~/build/amazon-freertos -DCOMPILER=cbmc
+-DBOARD=windows -DVENDOR=pc
+cmake --build ~/build/amazon-freertos --target all-proofs
+cd ~/build/amazon-freertos && ctest -L cbmc
+```
+
+Alternatively, this single command does the same thing, assuming you
+have the Ninja build tool installed:
+
+```sh
+ctest --build-and-test                \
+    ~/src/amazon-freertos             \
+    ~/build/amazon-freertos           \
+    --build-target cbmc               \
+    --build-generator Ninja           \
+    --build-options                   \
+      -DCOMPILER=cbmc                 \
+      -DBOARD=windows                 \
+      -DVENDOR=pc                     \
+    --test-command ctest -j4 -L cbmc --output-on-failure
+```
+
+
+
+Python-based build
+------------------
+
+### Prerequisites
 
 You will need Python 3. On Windows, you will need Visual Studio 2015 or later
 (in particular, you will need the Developer Command Prompt and NMake). On macOS
 and Linux, you will need Make.
 
 
-Installing CBMC
----------------
+### Installing CBMC
```

Comment:
These steps aren't working for me on native Windows. I do have win_bison.exe and win_flex.exe in the local directory.

C:\GitHub\cbmc>cmake -H. -Bbuild -DCMAKE_BUILD_TYPE=Release -DWITH_JBMC=OFF
-- Building for: Visual Studio 16 2019
-- Selecting Windows SDK version 10.0.18362.0 to target Windows 10.0.17763.
-- The C compiler identification is MSVC 19.23.28107.0
-- The CXX compiler identification is MSVC 19.23.28107.0
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.23.28105/bin/Hostx64/x64/cl.exe
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.23.28105/bin/Hostx64/x64/cl.exe -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.23.28105/bin/Hostx64/x64/cl.exe
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.23.28105/bin/Hostx64/x64/cl.exe -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found Git: C:/Program Files/Git/cmd/git.exe (found version "2.14.2.windows.1")
-- Could NOT find BISON (missing: BISON_EXECUTABLE)
-- Could NOT find FLEX (missing: FLEX_EXECUTABLE)
-- Could NOT find Doxygen (missing: DOXYGEN_EXECUTABLE)
CMake Error at src/CMakeLists.txt:26 (bison_target):
  Unknown CMake command "bison_target".
Call Stack (most recent call first):
  src/ansi-c/CMakeLists.txt:1 (generic_bison)
-- Configuring incomplete, errors occurred!
See also "C:/GitHub/cbmc/build/CMakeFiles/CMakeOutput.log".

Reply:
Here's the command-line that works for me on Windows:

`cmake -H. -Bbuild -DCMAKE_BUILD_TYPE=Release -DWITH_JBMC=OFF -DBISON_EXECUTABLE=c:\GitHub\cbmc\win_bison.exe -DFLEX_EXECUTABLE=c:\GitHub\cbmc\win_flex.exe`

Reply:
More progress, per offline discussion, now using Python 3.8.1 on Windows:

C:\GitHub\afr\tools\cbmc\proofs>python prepare.py
patch C:\GitHub\afr\tools\cbmc\patches\0001-Remove-static-storage-class-from-entry-points.patch
patch C:\GitHub\afr\tools\cbmc\patches\0002-Change-FreeRTOS_IP_Private.h-union-to-struct.patch
patch C:\GitHub\afr\tools\cbmc\patches\0005-remove-static-from-prvCopyDataToQueue.patch
error: patch failed: freertos_kernel/queue.c:105
error: freertos_kernel/queue.c: patch does not apply
prepare.py: ERROR patching failed: C:\GitHub\afr\tools\cbmc\patches\0005-remove-static-from-prvCopyDataToQueue.patch
patch C:\GitHub\afr\tools\cbmc\patches\0005-Remove-volatile-qualifier-from-tasks-variables.patch
error: patch failed: freertos_kernel/tasks.c:331
error: freertos_kernel/tasks.c: patch does not apply
prepare.py: ERROR patching failed: C:\GitHub\afr\tools\cbmc\patches\0005-Remove-volatile-qualifier-from-tasks-variables.patch
patch C:\GitHub\afr\tools\cbmc\patches\0006-Remove-static-from-prvNotifyQueueSetContainer.patch
error: patch failed: freertos_kernel/queue.c:207
error: freertos_kernel/queue.c: patch does not apply
prepare.py: ERROR patching failed: C:\GitHub\afr\tools\cbmc\patches\0006-Remove-static-from-prvNotifyQueueSetContainer.patch
patch C:\GitHub\afr\tools\cbmc\patches\0007-Remove-static-from-prvUnlockQueue.patch
error: patch failed: freertos_kernel/queue.c:175
error: freertos_kernel/queue.c: patch does not apply
prepare.py: ERROR patching failed: C:\GitHub\afr\tools\cbmc\patches\0007-Remove-static-from-prvUnlockQueue.patch
patch C:\GitHub\afr\tools\cbmc\patches\0008-Remove-static-attributes-from-functions-implementing.patch